runtime.md

```diff
@@ -1,5 +1,47 @@
 # Runtime and Lifecycle
 
+## State
+
+The runtime state for a container is persisted on disk so that external tools can consume and act on this information.
+The runtime state is stored in a JSON encoded file.
+It is recommended that this file is stored in a temporary filesystem so that it can be removed on a system reboot.
+On Linux based systems the state information should be stored in `/run/oci/containers`.
+The directory structure for a container is `/run/oci/containers/<containerID>/state.json`.
+By providing a default location that container state is stored external applications can find all containers running on a system.
+
+* **version** (string) Version of the OCI specification used when creating the container.
+* **id** (string) ID is the container's ID.
+* **pid** (int) Pid is the ID of the main process within the container.
+* **root** (string) Root is the path to the container's bundle directory.
+
+The ID is provided in the state because hooks will be executed with the state as the payload.
+This allows the hook to perform clean and teardown logic after the runtime destroys it's own state.
+The root directory to the bundle is provided in the state so that consumers can find the container's configuration and rootfs.
+
+*Example*
+
+```json
+{
+    "id": "oci-container",
+    "pid": 4422,
+    "root": "/containers/redis"
+}
+```
+
+Linux systems add some platform specific information to the state.
+
+* **externalFds** Paths to the container's open file descriptors.
+
+*Example Linux*
+
+```json
+{
+    "externalFds": [
+        "/proc/33/fd/1"
```

Comment:
On Fri, Aug 28, 2015 at 02:53:56AM -0700, Andrew Vagin wrote:
> "lsof -p 4196 +E" enumerates all processes, so this operation can
> require dozen seconds on hosts with 20K processes. Any solutions
> which require to enumerate all tasks in procfs doesn't work, because
> a container should be frozen for this time.

Ah, I wasn't sure how it was working.  It makes sense that we might
want to cache information that would let us avoid that loop.  Although
it seems like the useful information to record would be what we know
about the other end of the file descriptor (e.g. “18132 pipe
4192,runc,3r 4192,runc,4w” in [1]), not information that can be
extracted directly from the application's PID (e.g. “pipe:[18132]” in
[1]).  But maybe the important information in externalFds is just the
order and fact that a given file descriptor was passed from the
launching runtime?

> lrwx------ 1 root root 64 Aug 28 12:01 /proc/2573/fd/385 -> socket:[3197260]
> lrwx------ 1 root root 64 Aug 28 12:01 /proc/2573/fd/389 -> socket:[3197260]
> lrwx------ 1 root root 64 Aug 28 12:02 /proc/2573/fd/405 -> socket:[3197298]
>
> "externalFds" : [" socket:[3197260]"]
>
> We create a socket, set it as a third descriptor for a process. Then
> a process creates one more socket for it's own needs. /monitor.sock
> is unreachable from a container.
> 
> Then we decided to dump this container. CRIU will check that it know
> how to restore all sockets. It's possible to restore
> socket:[3197298], because it has been created inside CT, but
> socket:[3197260] should be marked as external. Currently we get this
> information from externalFds.
> 
> Then we prepare a new socket and restore a container. All
> descriptors which were linked with "socket:[3197260]" will be linked
> with the new socket.

This is the tricky part.  How does the restoring application know to
create that socket?  For things like #113's additional file
descriptors, you may have access to some user-supplied sockets, etc.,
and you may also have some runtime-created pipes that you intend to
connect to the application's standard streams.  How do you know which
of those should replace socket:[3197260]?  I'd rather see something
that tagged the supplied objects:

  "externalFds": {
    "stdin": "pipe:[18132]",
		"stdout": "pipe:[18133]",
		"stderr": "pipe:[18134]",
		"socket": "socket:[3197298]"
  }

where the standard stream keys would default to being created by the
runtime (when process.terminal was false), but all keys could be set
by the runtime-caller on invocation.  For example, if you were passing
a systemd-created socket through to runc on fd 8, you'd use:

  runc start --file-descriptor socket:8

and runc would inspect file descriptor 8, see that it was
socket:[3197298], and write that entry in externalFds.

On restore time, maybe you were passing the systemd-created socket
through to runc on fd 3, and you'd use:

  runc restore --file-descriptor socket:4

and whatever you passed in would be wired up where socket:[3197298]
had been on the checkpointed container.

[1]: https://github.com/opencontainers/specs/pull/87#discussion_r38173424